spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/security/IapAuthenticationAutoConfiguration.java

```diff
@@ -16,59 +16,142 @@
+ *         audience is provided through {@code spring.cloud.gcp.security.iap.audience} property)
+ *     </ul>
  * </ul>
- * <p>If these conditions are met, a custom {@link BearerTokenResolver} and an ES256 web registry-based JWT token
- * decoder beans are provided.
  * <p>If a custom {@link WebSecurityConfigurerAdapter} is present, it must add {@code .oauth2ResourceServer().jwt()}
  * customization to {@link org.springframework.security.config.annotation.web.builders.HttpSecurity} object. If no
  * custom {@link WebSecurityConfigurerAdapter} is found,
  * Spring Boot's default {@code OAuth2ResourceServerWebSecurityConfiguration} will add this customization.
  *
  * @author Elena Felder
+ *
  * @since 1.1
  */
 @Configuration
-@ConditionalOnProperty("spring.cloud.gcp.security.iap.enabled")
-@ConditionalOnClass({NimbusJwtDecoderJwkSupport.class})
+@ConditionalOnProperty(value = "spring.cloud.gcp.security.iap.enabled", matchIfMissing = true)
+@ConditionalOnClass({AudienceValidator.class})
 @AutoConfigureBefore(OAuth2ResourceServerAutoConfiguration.class)
+@AutoConfigureAfter(GcpContextAutoConfiguration.class)
 @EnableConfigurationProperties(IapAuthenticationProperties.class)
 public class IapAuthenticationAutoConfiguration {
 
+	private static final Log LOGGER = LogFactory.getLog(IapAuthenticationAutoConfiguration.class);
+
 	@Bean
 	@ConditionalOnMissingBean
-	public JwtDecoder iapJwtDecoder(IapAuthenticationProperties properties) {
-		NimbusJwtDecoderJwkSupport jwkSupport
-				= new NimbusJwtDecoderJwkSupport(properties.getRegistry(), properties.getAlgorithm());
-		jwkSupport.setJwtValidator(JwtValidators.createDefaultWithIssuer(properties.getIssuer()));
-		return jwkSupport;
+	public BearerTokenResolver iatTokenResolver(IapAuthenticationProperties properties) {
+		return r -> r.getHeader(properties.getHeader());
 	}
 
 	@Bean
 	@ConditionalOnMissingBean
-	public BearerTokenResolver iatTokenResolver(IapAuthenticationProperties properties) {
-		return r -> r.getHeader(properties.getHeader());
+	@ConditionalOnProperty("spring.cloud.gcp.security.iap.audience")
+	public AudienceValidator propertyBasedAudienceValidator(IapAuthenticationProperties properties) {
+		return new AudienceValidator(properties::getAudience);
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	@ConditionalOnGcpEnvironment(GcpEnvironment.ANY_APP_ENGINE)
+	public AudienceProvider appEngineBasedAudienceProvider(GcpProjectIdProvider projectIdProvider) {
+		return new AppEngineAudienceProvider(projectIdProvider);
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	@ConditionalOnGcpEnvironment(GcpEnvironment.GKE_OR_GCE)
+	public AudienceProvider computeEngineBasedAudienceProvider(GcpProjectIdProvider projectIdProvider) {
+		return new ComputeEngineAudienceProvider(projectIdProvider);
+	}
+
+	@Bean
+	@ConditionalOnBean(AudienceProvider.class)
+	public AudienceValidator environmentBasedAudienceValidator(AudienceProvider audienceProvider) {
```

Comment:
Yes, good catch.
I might be able to suppress AudienceValidator creation when testing the specific audience providers, so that they are never tested together (and so the hard-to-override functionality is never triggered). I'll try that.